source/plugin/event/causes.rst

```diff
@@ -69,55 +97,137 @@ object matching the provided type.
+    Some common identifying names for ``EventContextKey``\s are present as static fields in the
+    ``EventContextKeys`` class.
 
 Creating custom Causes
 ~~~~~~~~~~~~~~~~~~~~~~
 
-Creating a cause to use when firing an event is extremely easy. The hardest part is deciding
-what information to include in the cause. If you're firing an event from your plugin which is
-usually triggered through other means perhaps you want to include your plugin container so
-other plugins know that the event comes from your plugin. Or if you are firing the event on
-behalf of a player due to some action it's usually a good idea to include that player in
-the cause.
+Creating a cause is easy, but depends on whether you are creating your cause on the main server
+thread or async.
 
 .. note::
 
     Cause objects are immutable therefore cannot be modified once created.
 
-Using :javadoc:`Cause#of(NamedCause)`, you can construct a cause from a series of objects. The objects will be added to
-the cause chain in the order that they are passed to the method, so the first object parameter will become the root
-cause. Remember that a ``Cause`` may not be empty, so at least one non-null parameter is always required.
+Using the CauseStackManager
+===========================
+
+.. note::
+
+    The ``CauseStackManager`` only works on the main server thread. If you call it from a
+    different thread, an ``IllegalStateException`` will be thrown. Ensure you are on the main
+    server thread **before** calling methods on the ``CauseStackManager``.
+
+If you are creating your event on the main thread, then use the :javadoc:`CauseStackManager`, which can
+be found at :javadoc:`Sponge#getCauseStackManager()`. The ``CauseStackManager`` tracks the potential
+causes of events as the game runs, allowing for easy retrieval of the current ``Cause`` without effort.
+To see the current cause, call :javadoc:`CauseStackManager#getCurrentCause()`. You may notice that your
+plugin's :javadoc:`PluginContainer` is already in the returned ``Cause``, as plugins are one of the
+objects tracked by the manager. Using the ``CauseStackManager`` for creating causes removes the
+need for boilerplate-like code where you supply objects like your plugin container, you can concentrate
+on adding your own causes.
+
+Before adding your own causes, you should push a cause stack frame to the manager. Adding a frame acts
+as a saved state, when you are done with your causes, the removal of the frame returns the manager to
+its original state.
+
+For example, if you were to fire an event that was simulating another player in a sudo like command,
+you may want to add your acting player in the cause and the game profile of the player that you are
+simulating in the context (as the simulated player is not directly responsible for the event being fired.)
+
+**Creating a custom Cause with the CauseStackManager**
+
+.. code-block:: java
+
+    // In your code these would be populated
```

Comment:
Maybe move these outside of the code blocks to make it tranlateable?

Reply:
It might be difficult because ideally the comments are attached to a line of code, but I'll do my best to add a summary of what I'm trying to convey. I'm inclined to leave the lines in regardless of whether I do that to maximise clarity for those who can read the comments.